Refactor find package.py

Odilhao opened the pull request:
## Summary

Refactored the `parse_package_list` function in `find_package.py` to improve maintainability and added comprehensive test coverage.

### Changes Made

**Refactoring:**
- Replaced 40+ elif statements with organized data structures (prefix_removals, package_mappings, lowercase_packages, pattern transformations)
- Improved code readability and made adding new package transformations much easier
- Fixed pattern matching logic to be more selective and accurate
- All existing functionality preserved with no breaking changes

**Testing:**
- Added comprehensive test suite with 9 tests covering all transformation scenarios
- Included integration test with real-world data (191 packages)
- Added test documentation and easy-to-use test runner
- All tests pass successfully

### Benefits

- **Maintainable**: Adding new package transformations is now straightforward
- **Reliable**: Comprehensive test coverage ensures correct behavior
- **Documented**: Clear test documentation explains all transformation rules
- **Verified**: Integration test validates compatibility with production data

The refactored code handles the same package name transformations as before but with significantly better organization and maintainability.

Odilhao committed:
Refactor parse_package_list function for better maintainability

- Replace 40+ elif statements with organized data structures
- Group transformations into logical categories:
  - prefix_removals: Remove prefixes like 'python-'
  - package_mappings: Exact package name mappings
  - lowercase_packages: Packages that need to be lowercased
  - apply_pattern_transformations: Pattern-based transformations
- Improve code readability and make it easier to add new transformations
- Fix pattern matching logic to be more selective (e.g., only transform packages with specific characters)
- All existing functionality preserved - no breaking changes

Odilhao committed:
Add comprehensive test suite for parse_package_list function

- Create test/ directory with comprehensive test coverage
- Add test_parse_package_list.py with 8 unit tests covering:
  - Prefix removals (python- packages)
  - Exact package mappings (typing_extensions, galaxy_importer, etc.)
  - Lowercase packages (PyYAML, Django, etc.)
  - Pattern-based transformations (opentelemetry, poetry, etc.)
  - No transformation needed packages
  - Mixed transformation scenarios
  - Empty input and whitespace handling
- Add test_integration.py with real-world data test (191 packages)
- Add test/README.md with comprehensive documentation
- Add run_tests.py for easy test execution
- Update .gitignore to exclude __pycache__/
- All 9 tests pass successfully

Odilhao committed:
Add GitHub Actions workflow for find_package.py tests

- Automatically runs tests when find_package.py or test files are modified
- Tests against Python 3.11 and 3.12 for compatibility
- Uses path-based triggers to run only when relevant files change
- Includes both unit tests and integration test execution

Odilhao committed:
Convert test suite from unittest to pytest

- Convert test_parse_package_list.py from unittest class to pytest functions
- Convert test_integration.py from unittest to pytest format
- Replace self.assertEqual() with assert statements
- Update run_tests.py to use pytest instead of unittest
- Update GitHub Actions workflow to use pytest
- Add test/requirements-test.txt with pytest dependencies
- Update test/README.md with pytest instructions and examples
- All 9 tests pass successfully with pytest
- Improved test runner with better error handling and dependency checking
- Drop run_tests.py

JacobCallahan commented on run_tests.py:
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env python3
```
This file shouldn't be needed anymore now that you're using pytest

JacobCallahan commented on test/test_integration.py:
```diff
@@ -0,0 +1,268 @@
+import os
+
+# Add the parent directory to the path so we can import find_package
+sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
```
not a blocker, but since this is present in both test files, this could be added to a conftest.py file in the form of a session-scoped auto-use fixture.

Odilhao replied:
Moved the oneline to `conftest.py`

with

```
#!/usr/bin/env python3

import sys
import os
import pytest


@pytest.fixture(scope="session", autouse=True)
def setup_import_path():
    """Add the parent directory to sys.path so we can import find_package module."""
    parent_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    if parent_dir not in sys.path:
        sys.path.insert(0, parent_dir) 
```

Odilhao committed:
Address review comment: Add conftest.py for shared test setup

- Add test/conftest.py with session-scoped auto-use fixture for sys.path setup
- Remove duplicate sys.path.insert() code from both test files
- Update test/README.md to document conftest.py in test structure
- Follows pytest best practices for shared test configuration
- All 9 tests continue to pass

Addresses review comment from @JacobCallahan about consolidating
duplicate sys.path setup code into conftest.py

pcreech reviewed (approved):
Locally, the tests seem to result in success.  The base functionality of find_packages.py also appears to be retained.  LGTM from a general correctness standpoint.

decko commented on find_package.py:
```diff
@@ -6,68 +6,102 @@
+    }
+    
+    # Packages that need to be lowercased
+    lowercase_packages = {
```
There's any package name that needs to keep its uppercase characters? If isn't, why not simply ask for lowercase all package names and remove this dict here?

decko commented on find_package.py:
```diff
@@ -6,68 +6,102 @@
+    # Pattern-based transformations
+    def apply_pattern_transformations(name):
+        # OpenTelemetry packages: replace - with _
+        if name.startswith('opentelemetry'):
```
That dict idea could also be extended a little to allow rules like this:
{"opentelemetry-": {"rule": "starts_with", "from": "-", "to": "_"}}.

This could avoid all those ifs here.

Odilhao replied:
The main problem is that we have packages that need to keep the `-`, it's python packaging in the end of the day, everyone is doing it's own thing.

Odilhao replied:
I'm going to address this in another PR.
